fbgemm_gpu/src/sparse_ops/sparse_ops_gpu.cpp

```diff
@@ -64,39 +67,72 @@ class PackSegments : public torch::autograd::Function<PackSegments> {
       torch::autograd::AutogradContext* ctx,
       const Tensor& t_in,
       const Tensor& lengths,
-      const int64_t max_length) {
-    const int64_t total_length = t_in.contiguous().size(0);
+      const at::SymInt& max_length) {
+    const at::SymInt total_length = t_in.sym_size(0);
+
+    at::AutoDispatchBelowADInplaceOrView guard;
+
+    static auto custom_pack_segments_op =
+        torch::Dispatcher::singleton()
+            .findSchemaOrThrow("fbgemm::pack_segments", "")
+            .typed<at::Tensor(
+                const at::Tensor&, const at::Tensor&, const at::SymInt)>();
+
+    Tensor res = custom_pack_segments_op.call(t_in, lengths, max_length);
+
     ctx->saved_data["max_length"] = max_length;
     ctx->saved_data["total_length"] = total_length;
     ctx->save_for_backward({lengths});
 
-    // Run the forward pass.
-    const auto& res = pack_segments_forward_cuda(t_in, lengths, max_length);
-
-    torch::autograd::variable_list outputs(1);
-    outputs[0] = res;
-    return outputs;
+    return {res};
   }
 
   static torch::autograd::variable_list backward(
       torch::autograd::AutogradContext* ctx,
       torch::autograd::variable_list grad_output) {
     TORCH_CHECK(grad_output.size() == 2 or grad_output.size() == 1);
     const Tensor& grad = grad_output[0];
-    const auto& max_length = ctx->saved_data["max_length"].toInt();
-    const auto& total_length = ctx->saved_data["total_length"].toInt();
+    const auto& max_length = ctx->saved_data["max_length"].toSymInt();
+    const auto& total_length = ctx->saved_data["total_length"].toSymInt();
 
     // Retrieve saved variables for backward.
     const auto& saved_variables = ctx->get_saved_variables();
     const auto& lengths = saved_variables[0];
 
     torch::autograd::variable_list grad_inputs(5);
+
+    static auto custom_unpack_segments_op =
+        torch::Dispatcher::singleton()
+            .findSchemaOrThrow("fbgemm::unpack_segments", "")
+            .typed<at::Tensor(
+                const at::Tensor&,
+                const at::Tensor&,
+                const at::SymInt,
+                const at::SymInt)>();
+
     grad_inputs[0] =
-        pack_segments_backward_cuda(grad, lengths, total_length, max_length);
+        custom_unpack_segments_op.call(grad, lengths, total_length, max_length);
     return grad_inputs;
   }
 };
 
+Tensor pack_segments_cpu(
+    const Tensor& t_in,
+    const Tensor& lengths,
+    const at::SymInt max_length) {
+  const auto& res = PackSegments::apply(t_in, lengths, max_length);
```

Comment:
You're registering this function directly to the `CPU` key, so you don't want to call `PackSegments::apply()` here - you can just call `pack_segments_forward_cpu`

In fact, I think this function is unused in this file?